### PR TITLE
Upgrade Prettier

### PR DESCRIPTION
**Pull request by molly-moen**
Upgrade to the latest version of Prettier (2.8.7). This is necessary so that we can add typescript linting to our repo. The main changes applied here are:
- add a space between function and (), so function() becomes function ()
- some changes to the way multi-line statements were broken up. Details under “Improved method chain breaking heuristic” [here](https://prettier.io/blog/2020/03/21/2.0.0)
- if we use a number as a key in an object, it removed quotes around the number. This is the only one I could see causing bugs.

I opted out of the following default changes in this upgrade (see `.prettierrc.js`)
- A trailing comma at the end of every object. Ex {name: 'value', name2: 'value2',}  (note comma after value2). Here’s their[ reasoning](https://medium.com/@nikgraf/why-you-should-enforce-dangling-commas-for-multiline-statements-d034c98e36f8)
- Always wrap arrow function arguments with parentheses. I.e. go from const fn = x => … to const fn = (x) => …. I kept this as is for now (no parens) but I am open to thoughts on this one.

## Links
- [Prettier upgrade documentation](https://prettier.io/blog/2020/03/21/2.0.0)
- jira ticket: [SL-692](https://codedotorg.atlassian.net/browse/SL-692)

## Testing story
I am relying on drone to validate this change. I don't see any issues with the spacing changes. I spot tested the number changes and saw no issues, although I added one ignore [here](https://github.com/code-dot-org/code-dot-org/pull/51065/files#diff-a689cab7ac267ecb38d14dc286cfc00dc25c8cff39c5ce78f58e6404739c7cabR11) because it was applying the rule inconsistently.

If you want to view the changes, I would highly recommend viewing with whitespace hidden. I looked at all the non-test file changes with whitespace hidden and it took me quite a while. I spot checked some of the test file changes, but since they are all run by drone they felt less necessary to validate manually.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Review by levadadenys — APPROVED** [review]
Wow, well done! Like those rules improvements! LGTM!

(As for opted out changes):
+1 for trailing commas in the end of objects.

And just of curiosity, what's the pros of `Always wrap arrow function arguments with parentheses. I.e. go from const fn = x => … to const fn = (x) => …. ` ? (IMO `const fn = x => …` is a little bit cleaner and more common)

**Comment by megcrenshaw** [comment]
My first job had the trailing comma rule, and it was an absolute pleasure. I would love to do this in our repo. But I do think if we enable it, doing it as a separate PR makes sense –– there are already so many files changed here.

I'm also pro keeping the parens for `const f = (x) => ` for similar reasons –– slightly easier to update, and sounds like it will make working in TS easier. But again, it could be a separate PR.

I was visually surprised by the `function ()` syntax (with the space) but understand [their reasoning](https://prettier.io/blog/2020/03/21/2.0.0#always-add-a-space-after-the-function-keyword-3903httpsgithubcomprettierprettierpull3903-by-j-f1httpsgithubcomj-f1-josephfrazierhttpsgithubcomjosephfrazier-sosukesuzukihttpsgithubcomsosukesuzuki-thorn0httpsgithubcomthorn0-7516httpsgithubcomprettierprettierpull7516-by-bakkothttpsgithubcombakkot).

In short, I'm a fan of all the defaults! Thanks for your work on this.

As far as test coverage, I agree with your statement

> if we use a number as a key in an object, it removed quotes around the number. This is the only one I could see causing bugs.

In that case, thoughts about disabling this rule for now and enabling it in a separate PR? It'd be easier to (a) revert if needed, and (b) get eyes on exactly what is changing.

**Comment by molly-moen** [comment]
@megcrenshaw I agree we could discuss the other defaults and potentially do them as follow-ups. I feel ambivalent about the trailing commas, and lean towards the parens around single params.

For the quotes/numbers, I found all the cases where that was changed and either validated them locally or added an ignore if I could not validate. Do you think that's sufficient? Unfortunately the prettier configuration value we would apply in this case is [preserve](https://prettier.io/docs/en/options.html#quote-props), which would mean I would either need to manually track down the quote changes (and hope I didn't miss any), or start over.

**Comment by megcrenshaw** [comment]
> @megcrenshaw I agree we could discuss the other defaults and potentially do them as follow-ups. I feel ambivalent about the trailing commas, and lean towards the parens around single params.
> 
> For the quotes/numbers, I found all the cases where that was changed and either validated them locally or added an ignore if I could not validate. Do you think that's sufficient? Unfortunately the prettier configuration value we would apply in this case is [preserve](https://prettier.io/docs/en/options.html#quote-props), which would mean I would either need to manually track down the quote changes (and hope I didn't miss any), or start over.

Ah, I see re: `preserve`. Makes sense! Thanks 👍

**Review by sanchitmalhotra126 — APPROVED** [review]
LGTM! I'm good with the defaults you've chosen.

**Comment by sanchitmalhotra126** [comment]
> My first job had the trailing comma rule, and it was an absolute pleasure. I would love to do this in our repo. But I do think if we enable it, doing it as a separate PR makes sense –– there are already so many files changed here.

I've haven't used the trailing comma rule (and initially thought I'd be opposed to it), but would be curious to try it out if it makes for cleaner diffs! FWIW, I personally use a prettier plugin in VSCode to lint as I go, so I don't think i'd notice either way.

**Comment by molly-moen** [comment]
@levadadenys [here](https://prettier.io/blog/2020/03/21/2.0.0#change-default-value-for-arrowparens-to-always-7430httpsgithubcomprettierprettierpull7430-by-kachkaevhttpsgithubcomkachkaev) is their reasoning for the parens around arrow params.

**Comment by levadadenys** [comment]
> @levadadenys [here](https://prettier.io/blog/2020/03/21/2.0.0#change-default-value-for-arrowparens-to-always-7430httpsgithubcomprettierprettierpull7430-by-kachkaevhttpsgithubcomkachkaev) is their reasoning for the parens around arrow params.

Oh, I see now, thanks for sharing. Agree, sounds reasonable to me. It can make dev process slightly more comfortable
